CogniteSdk.Types/Beta/DataModels/Instances/InstanceWrite.cs

```diff
@@ -0,0 +1,196 @@
+        /// </summary>
+        edge
+    }
+
+    /// <summary>
+    /// Base class for ingestion of nodes or edges.
+    /// </summary>
+    public abstract class BaseInstanceWrite
+    {
+        /// <summary>
+        /// Type of instance.
+        /// </summary>
+        public InstanceType InstanceType { get; set; }
+        /// <summary>
+        /// Fail the ingestion request if the node's version is greater than or equal to this value.
+        /// If no existingVersion is specified, the ingestion will always overwrite any existing data.
+        /// If existingVersion is set to0, the upsert will behave as an insert, so it will fail if
+        /// the item already exists.
+        /// If skipOnVersionConflict is set on the request, then the item will be skipped instead of
+        /// failing.
+        /// </summary>
+        public int? ExistingVersion { get; set; }
+        /// <summary>
+        /// Id of the space that the node belongs to. This space-id cannot be updated.
+        /// </summary>
+        public string Space { get; set; }
+        /// <summary>
+        /// Unique identifier for the node.
+        /// </summary>
+        public string ExternalId { get; set; }
+        /// <summary>
+        /// List of source properties to write. The properties are from the
+        /// views/containers that make up this instance.
+        /// </summary>
+        public IEnumerable<InstanceData> Sources { get; set; }
+    }
+
+    /// <summary>
+    /// Ingest a node.
+    /// </summary>
+    public class NodeWrite : BaseInstanceWrite
+    {
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public NodeWrite()
+        {
+            InstanceType = InstanceType.node;
+        }
+    }
+
+    /// <summary>
+    /// Ingest an edge.
+    /// </summary>
+    public class EdgeWrite : BaseInstanceWrite
+    {
+        /// <summary>
+        /// Constructor
+        /// </summary>
+        public EdgeWrite()
+        {
+            InstanceType = InstanceType.edge;
+        }
+        /// <summary>
+        /// Edge type as direct relation to another node.
+        /// </summary>
+        public DirectRelationIdentifier Type { get; set; }
+        /// <summary>
+        /// Start node of this edge.
+        /// </summary>
+        public DirectRelationIdentifier StartNode { get; set; }
+        /// <summary>
+        /// End node of this edge.
+        /// </summary>
+        public DirectRelationIdentifier EndNode { get; set; }
+    }
+
+
+    /// <summary>
+    /// Base class for user-defined property value groups.
+    /// Use <see cref="InstanceData{T}"/> to set values, you can use
+    /// your own DTOs, dynamic, or <see cref="StandardInstanceWriteData"/> as contents,
+    /// for example.
+    /// </summary>
+    public abstract class InstanceData
+    {
+        /// <summary>
+        /// Source of the instance.
+        /// </summary>
+        public SourceIdentifier Source { get; set; }
```

Comment:
It's either a view or a container identifier. Container does not have version.

Reply:
Do we not handle the case where we have a view as source then? I was thinking this would need to be something similar to other union implementations in the PR.

Reply:
It is! `SourceIdentifier` is a tagged union type.

Reply:
Ah, I missed that one! Nice!